dali/operators/random/rng_base.h

```diff
@@ -0,0 +1,130 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_RANDOM_RNG_BASE_CPU_H_
+#define DALI_OPERATORS_RANDOM_RNG_BASE_CPU_H_
+
+#include <random>
+#include <vector>
+#include "dali/core/convert.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/pipeline/util/batch_rng.h"
+#include "dali/core/static_switch.h"
+
+namespace dali {
+
+DALI_SCHEMA(RNGAttr)
+    .DocStr(R"code()code")
+    .AddOptionalArg<std::vector<int>>("shape",
+      R"code(Shape of the data.)code", nullptr, true)
+    .AddOptionalArg<DALIDataType>("dtype",
+      R"code(Data type.)code", nullptr);
+
+template<typename Backend, typename Impl>
+class RNGBase : public Operator<Backend> {
+ public:
+  ~RNGBase() override = default;
+
+ protected:
+  explicit RNGBase(const OpSpec &spec)
+      : Operator<Backend>(spec),
+        rng_(spec.GetArgument<int64_t>("seed"), spec.GetArgument<DALIDataType>("batch_size")) {
+  }
+
+  Impl &This() noexcept { return static_cast<Impl&>(*this); }
+  const Impl &This() const noexcept { return static_cast<const Impl&>(*this); }
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc,
+                 const workspace_t<Backend> &ws) override {
+    if (!spec_.TryGetArgument(dtype_, "dtype"))
+      dtype_ = This().DefaultDataType();
+
+    bool has_shape = spec_.ArgumentDefined("shape");
+    bool has_shape_like = spec_.NumRegularInput() == 1;
+    DALI_ENFORCE(!(has_shape && has_shape_like),
+      "Providing argument \"shape\" is incompatible with providing a shape-like input");
+
+    if (has_shape_like) {
+      shape_ = ws.template InputRef<Backend>(0).shape();
+    } else if (has_shape) {
+      GetShapeArgument(shape_, spec_, "shape", ws);
+    } else {
+      shape_ = uniform_list_shape(spec_.template GetArgument<DALIDataType>("batch_size"), {1});
+    }
+    batch_size_ = shape_.size();
+
+    This().AcquireArgs(spec_, ws, batch_size_);
+
+    output_desc.resize(1);
+    output_desc[0].shape = shape_;
+    output_desc[0].type = TypeTable::GetTypeInfo(dtype_);
+    return true;
+  }
+
+  template <typename T>
+  void RunImplTyped(workspace_t<CPUBackend> &ws) {
+    auto &output = ws.OutputRef<CPUBackend>(0);
+    auto out_view = view<T>(output);
+    const auto &out_shape = out_view.shape;
+    auto &tp = ws.GetThreadPool();
+    constexpr int64_t kThreshold = 0; 1 << 18;
+    constexpr int64_t kChunkSize = 1 << 16;
+    constexpr size_t kNumChunkSeeds = 4;
```

Comment:
The internal state of the generator fits 623 32-bit integers. I think we can splurge and have 16 integers here.